anomdtct/anomdtct/pipeline.py

```diff
@@ -0,0 +1,151 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+from datetime import timedelta, date
+import pandas as pd
+from google.cloud.bigquery._helpers import _bytes_to_json
+from google.cloud import bigquery
+from google.cloud.exceptions import NotFound
+from anomdtct.utils import s2d
+from anomdtct.data import get_raw_data, prepare_data, get_data_for_date, prepare_training_data
+from anomdtct.forecast import forecast, fit_model
+from anomdtct.output import write_records, write_to_spreadsheet, write_model
+import logging
+
+
+DEFAULT_BQ_PROJECT = "moz-fx-data-shared-prod"
+DEFAULT_BQ_DATASET = "analysis"
+DEFAULT_BQ_TABLE = "deviations"
+
+DEFAULT_BQ_MODEL_CACHE_PROJECT = "moz-fx-data-shared-prod"
+DEFAULT_BQ_MODEL_CACHE_DATASET = "analysis"
+DEFAULT_BQ_MODEL_CACHE_TABLE = "deviations_model_cache"
+
+METRICS = {
+    "light_funnel_dau_city": "desktop_dau",
+    "light_funnel_dau_country": "desktop_dau",
+    "light_funnel_mean_active_hours_per_profile_city":
+        "mean_active_hours_per_client",
+    "light_funnel_mean_active_hours_per_profile_country":
+        "mean_active_hours_per_client",
+}
+
+
+def fit_models(    
+    bq_client,
+    bq_storage_client,
+    project_id=DEFAULT_BQ_MODEL_CACHE_PROJECT,
+    dataset_id=DEFAULT_BQ_MODEL_CACHE_DATASET,
+    table_id=DEFAULT_BQ_MODEL_CACHE_TABLE
+):
+    training_start_date = '2016-04-08'
+    training_end_date = '2020-01-30'
+
+    for metric in METRICS.keys():
+        # overwrite existing table for caching models
+        table = '.'.join([project_id, dataset_id, table_id])
+        bq_client.delete_table(table)
+
+        raw_data = get_raw_data(
+            bq_client,
+            bq_storage_client,
+            metric,
+            training_start_date,
+            training_end_date
+        )
+        clean_training_data = prepare_training_data(
+            raw_data, s2d(training_start_date), s2d(training_end_date)
+        )
+        for c in clean_training_data.keys():
+            if (len(clean_training_data[c]) < 600):
+                continue
+
+            pickled_model = fit_model(clean_training_data, c)
+
+            record = {
+                "metric": metric,
+                "geography": c,
+                "model": _bytes_to_json(pickled_model)
+            }
+
+            write_model(bq_client, table, record)
```

Comment:
How many models are we fitting in this loop, and how large are the pickled models? We may want to build up a list of these pickled models and save them all in one load job. We could have load jobs fail if we try to run many of them in quick succession.